Project2

GaxysGalactic opened the pull request:
Doubly Linked List, Stack, and Queue

GaxysGalactic committed:
Doubly Linked List

+ Added Doubly Linked List template
o Fixed a few things in Singly Linked

GaxysGalactic committed:
Stack & Queue

+ Added Stack template
+ Added Queue template

GaxysGalactic committed:
Bugfixes

o Fixed some bugs in Remove and RemoveHead for Doubly Linked List

GaxysGalactic committed:
Bugfixes and Testing

o Changed the cpp to test the DLList instead
o Made a few bugfixes related to implementation

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+	//-------------------------------------------------------------------------------------------
+	TDLList()
+	{
+
```
could you do it in the same line :3 (empty space unnecessary)

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+		{
+			AddTail(NewData);
+		}
+		else if (Index < Size)
```
in this case, it is unnecessary to declare the if statement. since you have already defined the previous conditions, it would be enough only with "else", because the program would already know

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+			FNode* Current = GetNode(Index);
+			Current->Previous->Next = Current->Next;
+
+			if (Index == Size - 1)
```
could you put `Current == Tail` it´s more readable but it's fine

(you can correct me if I'm wrong)

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//										ITERATOR
```
maybe you should put this closer to the class FIterator, just to sort your class, furthermore, you could make your class more readable and you can have all elements related to the iterator in the same code chunk

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+	//-------------------------------------------------------------------------------------------
+	//										UTILITY FUNCTIONS
+	//-------------------------------------------------------------------------------------------
+	FNode* GetNode(const int Index)
```
maybe you should have a const function

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/Queue.h:
```diff
@@ -0,0 +1,47 @@
+		return List.GetHead();
+	}
+
+	bool IsEmpty()
```
this function should be a const function

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/Queue.h:
```diff
@@ -0,0 +1,47 @@
+		return List.IsEmpty();
+	}
+
+	int GetSize()
```
you should have a const function to this

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/SLList.h:
```diff
@@ -230,6 +229,20 @@ class TSLList
+		}
+		--Size;
+	}
+
```
LOL! I can´t comment from this part down, btw, just a comment about the clear method, you should call RemoveHead() inside this function, maybe it is easier to read since you created the method.

Sergio6040 replied:
I agree 👍

simonkytr commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/Stack.h:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool IsEmpty()
```
it maybe could be const function

simonkytr reviewed (changes requested):
excellent job, but I've made a few comments, it's nothing unusual

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/Stack.h:
```diff
@@ -0,0 +1,55 @@
+	int GetSize()
+	{
+		return List.GetSize();
+	}
```
also this should be const

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/Stack.h:
```diff
@@ -0,0 +1,55 @@
+	TStack()
+	{
+
+	}
```
if you are not creating your own constructor, the default will be created automatically, this is not necessary

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/SLList.cpp:
```diff
@@ -4,10 +4,10 @@
 #include <iostream>
 #include <string>
 
```
this file should be called, Main or Source, anything but SLList.cpp, because it is confusing

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+	T& operator[](const int Index)
+	{
+		FNode* Node = GetNode(Index);
+		return Node->Data;
```
here you can use, GetNode(Index)->Data;

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+	{
+		FNode* Node = new FNode;
+		Node->Data = NewData;
+		Node->Next = Head;
```
you can create a constructor for the FNode, where you can do this 👍

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+		delete Head;
+		Head = NewHead;
+
+		if (Size == 1)
+		{
+			Tail = nullptr;
+		}
+		else
+		{
+			Head->Previous = nullptr;
+		}
```
I think this is not rigth, if size == 1 Tail will be = Head, Doesn't it?

and Head->Previos = nullptr should always be called

also xd better if you do a bounce checking before anything

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/DLList.h:
```diff
@@ -0,0 +1,398 @@
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		while (Head)
+		{
+			RemoveHead();
+		}
+	}
```
why you and david have the same code, this looks suspicious 👀

Sergio6040 commented on Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/Queue.h:
```diff
@@ -0,0 +1,47 @@
+
+public:
+
+	TQueue()
+	{
+
+	}
```
here you can use 
```
TQueue() = Default;
```

GaxysGalactic committed:
Code Review Changes

+ Added a constructor for FNode for efficiency
+ Added separators in Queue.h
o Fixed some const issues
o Renamed SLList.cpp -> Main.cpp
- Removed default constructors from Stack.h and Queue.h, as those are now automatic, since I have no other constructors yet.
